lib/DxilPIXPasses/DxilDebugInstrumentation.cpp

```diff
@@ -850,99 +986,199 @@ void DxilDebugInstrumentation::addStepEntryForType(
+        ret.ConstantAllocaStoreValue = dyn_cast<ConstantFP>(ValueAsConst)
+                                           ->getValueAPF()
+                                           .bitcastToAPInt()
+                                           .getLimitedValue();
+        break;
+      case Type::IntegerTyID:
+        ret.ConstantAllocaStoreValue =
+            dyn_cast<ConstantInt>(ValueAsConst)->getLimitedValue();
+        break;
+      default:
+        return std::nullopt;
+      }
+      return ret;
+    }
+  }
+  return std::nullopt;
 }
 
-void DxilDebugInstrumentation::addStepDebugEntry(BuilderContext &BC,
-                                                 Instruction *Inst) {
-  if (Inst->getOpcode() == Instruction::OtherOps::PHI) {
-    return;
-  }
+std::optional<InstructionAndType>
+DxilDebugInstrumentation::addStepDebugEntry(BuilderContext *BC,
+                                            Instruction *Inst) {
   if (PIXPassHelpers::IsAllocateRayQueryInstruction(Inst)) {
-    return;
+    return std::nullopt;
   }
 
   if (auto *St = llvm::dyn_cast<llvm::StoreInst>(Inst)) {
-    addStoreStepDebugEntry(BC, St);
-    return;
+    return addStoreStepDebugEntry(BC, St);
   }
 
   std::uint32_t InstNum;
   if (!pix_dxil::PixDxilInstNum::FromInst(Inst, &InstNum)) {
-    return;
+    return std::nullopt;
+  }
+
+  if (auto *Ld = llvm::dyn_cast<llvm::LoadInst>(Inst)) {
+    if (llvm::isa<ConstantExpr>(Ld->getPointerOperand())) {
+      auto *constant = llvm::cast<ConstantExpr>(Ld->getPointerOperand());
+      if (constant->getOpcode() == Instruction::GetElementPtr) {
+        PIXPassHelpers::ScopedInstruction asInstr(constant->getAsInstruction());
+        auto *GEP = llvm::cast<GetElementPtrInst>(asInstr.Get());
+        if (GEP->getPointerOperand()->getName().equals("dx.nothing.a")) {
+          // These debug-only loads are interesting as instructions to
+          // step though where otherwise no step might exist for the
+          // given HLSL lines, so we include them in the instrumentation:
+          InstructionAndType ret{};
+          ret.Inst = Inst;
+          ret.InstructionOrdinal = InstNum;
+          ret.Type = DebugShaderModifierRecordTypeDXILStepVoid;
+          return ret;
+        }
+      }
+    }
   }
 
   std::uint32_t RegNum;
   if (!pix_dxil::PixDxilReg::FromInst(Inst, &RegNum)) {
-    if (Inst->getOpcode() == Instruction::Ret)
-      addStepEntryForType<void>(DebugShaderModifierRecordTypeDXILStepTerminator,
-                                BC, InstNum, nullptr, 0, 0);
-    return;
+    if (Inst->getOpcode() == Instruction::Ret) {
+      if (BC != nullptr)
+        addStepEntryForType<void>(DebugShaderModifierRecordTypeDXILStepRet, *BC,
+                                  InstNum, nullptr, 0, 0);
+      InstructionAndType ret{};
+      ret.Inst = Inst;
+      ret.InstructionOrdinal = InstNum;
+      ret.Type = DebugShaderModifierRecordTypeDXILStepRet;
+      return ret;
+    } else if (Inst->isTerminator()) {
```

Comment:
It's a very late nitpick, but just as an FYI, Instruction::Ret is a terminator and will be true here as well. You might have checked terminator first and included all of the common code and only checked for the ret opcode for the last type assignment.

Reply:
Thanks. Yes, this could be telescoped a teeny bit.

Reply:
Made a note to fix this in another checkin.